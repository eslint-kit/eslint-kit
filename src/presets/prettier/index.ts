import { Config } from 'prettier'
import { publicPresetNames } from '../names'
import { createPreset } from '../shared'

export const prettier = createPreset<Config>({
  name: publicPresetNames.prettier,
  compile: ({ options }) => ({
    plugins: ['prettier'],
    rules: {
      'prettier/prettier': options ? ['warn', options] : 'warn',
    },
  }),
})
