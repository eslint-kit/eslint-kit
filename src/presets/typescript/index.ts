import path from 'path'
import { Linter } from '@typescript-eslint/utils/dist/ts-eslint'
import { EXTENSIONS, importExtensionsRule } from '../../shared'
import { createPreset } from '../shared'

export interface Options {
  root?: string
  tsconfig?: string
}

export const typescriptRules: Linter.RulesRecord = {
  '@typescript-eslint/adjacent-overload-signatures': 'warn',
  '@typescript-eslint/ban-ts-comment': 'error',
  '@typescript-eslint/ban-types': 'error',
  '@typescript-eslint/naming-convention': [
    'warn',
    {
      selector: 'variable',
      format: ['camelCase', 'UPPER_CASE', 'PascalCase'],
      leadingUnderscore: 'allow',
    },
    {
      selector: 'parameter',
      format: ['camelCase'],
      leadingUnderscore: 'allow',
    },
    {
      selector: 'typeLike',
      format: ['PascalCase'],
    },
  ],
  '@typescript-eslint/consistent-type-assertions': 'warn',
  '@typescript-eslint/consistent-type-definitions': ['warn', 'interface'],
  '@typescript-eslint/no-array-constructor': 'warn',
  '@typescript-eslint/no-empty-interface': 'warn',
  '@typescript-eslint/no-explicit-any': 'warn',
  '@typescript-eslint/no-extra-non-null-assertion': 'warn',
  '@typescript-eslint/no-for-in-array': 'warn',
  '@typescript-eslint/no-inferrable-types': 'error',
  '@typescript-eslint/no-misused-new': 'error',
  '@typescript-eslint/no-namespace': 'error',
  '@typescript-eslint/no-non-null-assertion': 'warn',
  '@typescript-eslint/no-unused-vars': [
    'warn',
    {
      ignoreRestSiblings: true,
      varsIgnorePattern: '^_',
      argsIgnorePattern: '^_',
    },
  ],
  '@typescript-eslint/no-unnecessary-condition': 'warn',
  '@typescript-eslint/no-unused-expressions': 'warn',
  '@typescript-eslint/no-useless-constructor': 'warn',
  '@typescript-eslint/prefer-for-of': 'warn',
  '@typescript-eslint/prefer-includes': 'warn',
  '@typescript-eslint/prefer-namespace-keyword': 'error',
  '@typescript-eslint/prefer-optional-chain': 'warn',
  '@typescript-eslint/prefer-string-starts-ends-with': 'warn',
  '@typescript-eslint/require-array-sort-compare': 'error',
  '@typescript-eslint/return-await': ['error', 'in-try-catch'],
  '@typescript-eslint/triple-slash-reference': 'error',

  // Specific rules that are enabled using @typescript-eslint, but have analogues in common eslint
  'camelcase': 'off',
  'no-array-constructor': 'off',
  'no-unused-vars': 'off',
  'no-unused-expressions': 'off',
  'no-useless-constructor': 'off',
  'no-return-await': 'off',

  // Checked by Typescript - ts(2378)
  'getter-return': 'off',
  // Checked by Typescript - ts(2300)
  'no-dupe-args': 'off',
  // Checked by Typescript - ts(1117)
  'no-dupe-keys': 'off',
  // Checked by Typescript - ts(7027)
  'no-unreachable': 'off',
  // Checked by Typescript - ts(2367)
  'valid-typeof': 'off',
  // Checked by Typescript - ts(2588)
  'no-const-assign': 'off',
  // Checked by Typescript - ts(2588)
  'no-new-symbol': 'off',
  // Checked by Typescript - ts(2376)
  'no-this-before-super': 'off',
  // This is checked by Typescript using the option `strictNullChecks`.
  'no-undef': 'off',
  // This is already checked by Typescript.
  'no-dupe-class-members': 'off',
  // This is already checked by Typescript.
  'no-redeclare': 'off',

  // Checked by Typescript
  'import/no-extraneous-dependencies': 'off',
}

export const typescript = createPreset<'typescript', Options | void>({
  name: 'typescript',
  updateMeta: ({ meta, options }) => {
    meta.extensions = meta.extensions.concat(EXTENSIONS.TS)
    meta.typescript.used = true
    meta.typescript.root = path.resolve(meta.root, options?.root ?? './')
    meta.typescript.tsconfig = options?.tsconfig ?? 'tsconfig.json'
  },
  compile: ({ meta }) => ({
    plugins: ['import', '@typescript-eslint'],
    parser: '@typescript-eslint/parser',
    parserOptions: {
      project: meta.typescript.tsconfig,
      createDefaultProgram: true,
      tsconfigRootDir: meta.typescript.root,
    },
    settings: {
      'import/extensions': meta.extensions,
      'import/resolver': {
        node: {
          extensions: meta.extensions,
        },
      },
      'import/ignore': ['\\.(coffee|scss|css|less|hbs|svg|json)$'],
      'import/external-module-folders': ['node_modules', 'node_modules/@types'],
      'import/parsers': {
        '@typescript-eslint/parser': EXTENSIONS.TS,
      },
    },
    rules: {
      'import/extensions': [
        importExtensionsRule[0],
        importExtensionsRule[1],
        meta.extensions.reduce<{ [key: string]: 'never' }>((acc, extension) => {
          const extensionShort = extension.slice(1)
          acc[extensionShort] = 'never'
          return acc
        }, {}),
      ],
    },
    overrides: [
      {
        files: ['*.ts', '*.tsx'],
        rules: typescriptRules,
      },
    ],
  }),
})
