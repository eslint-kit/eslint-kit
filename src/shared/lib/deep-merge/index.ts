/* eslint-disable @typescript-eslint/no-explicit-any */
import { Merge, PathsOf } from './types'

function isObject(value: any): value is Record<string, any> {
  return typeof value === 'object' && value !== null
}

export enum Strategy {
  Shallow,
  Deep,
  Override,
}

type AllPaths<A, B> = PathsOf<A> | PathsOf<B> | null

function shallowMerge<A, B>(a: A, b: B): Merge<A, B> {
  let merged: any

  if (Array.isArray(a) && Array.isArray(b)) {
    merged = [...a, ...b]
  } else if (isObject(a) && isObject(b)) {
    merged = { ...a, ...b }
  } else if (typeof b !== 'undefined') {
    merged = b
  } else {
    merged = a
  }

  return merged as Merge<A, B>
}

// eslint-disable-next-line sonarjs/cognitive-complexity
export function deepMerge<A, B>(
  a: A,
  b: B,
  strategy: (path: AllPaths<A, B>) => Strategy = () => Strategy.Deep,
  basePath: AllPaths<A, B> = null,
): Merge<A, B> {
  if (strategy(basePath) === Strategy.Shallow) {
    return shallowMerge(a, b)
  }

  if (strategy(basePath) === Strategy.Override) {
    return b as Merge<A, B>
  }

  if (Array.isArray(a) && Array.isArray(b)) {
    return shallowMerge(a, b)
  }

  if (isObject(a) && isObject(b)) {
    const keys = [...Object.keys(a), ...Object.keys(b)]
    const final: any = {}

    for (const key of keys) {
      const path = basePath ? [basePath, key].join('.') : key
      final[key] = deepMerge(
        (a as any)[key],
        (b as any)[key],
        strategy,
        path as AllPaths<A, B>,
      )
    }

    return final
  }

  return shallowMerge(a, b)
}
