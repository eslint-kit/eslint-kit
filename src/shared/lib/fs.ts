import fs from 'fs'
import path from 'path'

export function readJson<T>(root: string, file: string): T | null {
  const fullpath = path.resolve(root, file)
  if (!fs.existsSync(fullpath)) return null
  const data = fs.readFileSync(fullpath, 'utf-8')
  try {
    return JSON.parse(data)
  } catch {
    return null
  }
}
