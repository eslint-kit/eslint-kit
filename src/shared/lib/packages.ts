import { parse } from 'semver'
import { PackageJson } from '../types'

function mergeDependencies(packageJson: PackageJson) {
  const { dependencies = {}, devDependencies = {} } = packageJson
  return { ...dependencies, ...devDependencies }
}

export function getPackageSemver(name: string, packageJson: PackageJson) {
  const allDependencies = mergeDependencies(packageJson)
  const version = allDependencies[name]
  if (!version) return null
  const parsed = parse(version)
  if (!parsed) return null
  return parsed
}
