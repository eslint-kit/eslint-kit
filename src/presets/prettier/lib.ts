import { Linter } from 'eslint'
import { Options } from 'prettier'
import { Meta } from '../shared'

export interface Input {
  options?: Options
  meta: Meta
}

export function createPrettierRule({ meta, options }: Input) {
  const plugins: string[] = []

  if (meta.imports.layout) {
    plugins.push('prettier-plugin-layout')
  }

  return [
    'warn',
    {
      plugins,
      ...options,
    },
  ] as Linter.RuleEntry
}
