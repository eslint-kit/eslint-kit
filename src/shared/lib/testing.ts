import fs from 'fs'
import path from 'path'
import { ESLint } from 'eslint'
import { Preset, PresetName, PRIORITY, compilePresets } from '../../presets'
import { base } from '../../presets/base'

export interface Options {
  presets: Preset[]
  priority?: PresetName[]
  dirname: string
  files: string[]
  extension: 'js' | 'jsx' | 'ts' | 'tsx' | 'svelte' | 'vue'
}

export async function testConfig({
  presets,
  priority = PRIORITY,
  dirname,
  files,
  extension,
}: Options) {
  const root = path.resolve(dirname, './tests')

  const cli = new ESLint({
    baseConfig: compilePresets([base({ root }), ...presets], priority),
    useEslintrc: false,
    cwd: root,
    ignore: false,
  })

  for (const file of files) {
    const filePath = path.resolve(root, `./${file}.${extension}`)
    const code = fs.readFileSync(filePath).toString()
    const result = await cli.lintText(code, { filePath })
    expect(result).toMatchSnapshot(`${file}`)
  }
}
