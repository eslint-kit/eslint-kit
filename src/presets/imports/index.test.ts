import { testConfig } from '../../shared/lib/testing'
import { prettier } from '../prettier'
import { typescript } from '../typescript'
import { imports } from './index'

describe('[Presets] Imports', () => {
  it('should match snapshots', async () => {
    await testConfig({
      presets: [imports()],
      dirname: __dirname,
      files: ['basic-clear', 'basic-warn', 'basic-error'],
      extension: 'js',
    })
  })

  it('should respect newline=true', async () => {
    await testConfig({
      presets: [imports({ sort: { newline: true } })],
      dirname: __dirname,
      files: ['newline-true'],
      extension: 'js',
    })
  })

  it('should respect newline=false', async () => {
    await testConfig({
      presets: [imports({ sort: { newline: false } })],
      dirname: __dirname,
      files: ['newline-false'],
      extension: 'js',
    })
  })

  it('should respect paths from jsconfig.json', async () => {
    await testConfig({
      presets: [imports()],
      dirname: __dirname,
      files: ['alias-jsconfig-clear', 'alias-jsconfig-error'],
      extension: 'js',
      disableLogs: true,
    })
  })

  it('should respect paths from tsconfig.json', async () => {
    await testConfig({
      presets: [typescript(), imports()],
      dirname: __dirname,
      files: ['alias-tsconfig-clear', 'alias-tsconfig-error'],
      extension: 'ts',
      disableLogs: true,
    })
  })

  it('should arrange imports in the form of a grid', async () => {
    await testConfig({
      presets: [imports({ layout: true }), prettier()],
      dirname: __dirname,
      files: ['import-layout-clear', 'import-layout-error'],
      extension: 'js',
      disableLogs: true,
    })
  })
})
