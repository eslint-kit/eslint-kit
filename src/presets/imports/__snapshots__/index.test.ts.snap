// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Presets] Imports should arrange imports in the form of a grid: import-layout-clear 1`] = `
Array [
  Object {
    "errorCount": 3,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/imports/tests/import-layout-clear.js",
    "fixableErrorCount": 0,
    "fixableWarningCount": 2,
    "messages": Array [
      Object {
        "column": 10,
        "endColumn": 21,
        "endLine": 1,
        "line": 1,
        "message": "'conditional' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "no-unused-vars",
        "severity": 1,
      },
      Object {
        "column": 24,
        "endColumn": 30,
        "endLine": 1,
        "fix": Object {
          "range": Array [
            23,
            29,
          ],
          "text": "",
        },
        "line": 1,
        "message": "Delete \`······\`",
        "messageId": "delete",
        "nodeType": null,
        "ruleId": "prettier/prettier",
        "severity": 1,
      },
      Object {
        "column": 35,
        "endColumn": 60,
        "endLine": 1,
        "line": 1,
        "message": "Unable to resolve path to module '../../shared/lib/eslint'.",
        "nodeType": "Literal",
        "ruleId": "import/no-unresolved",
        "severity": 2,
      },
      Object {
        "column": 35,
        "endColumn": 60,
        "endLine": 1,
        "line": 1,
        "message": "Missing file extension for \\"../../shared/lib/eslint\\"",
        "nodeType": "Literal",
        "ruleId": "import/extensions",
        "severity": 1,
      },
      Object {
        "column": 10,
        "endColumn": 27,
        "endLine": 2,
        "line": 2,
        "message": "'publicPresetNames' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "no-unused-vars",
        "severity": 1,
      },
      Object {
        "column": 35,
        "endColumn": 45,
        "endLine": 2,
        "line": 2,
        "message": "Unable to resolve path to module '../names'.",
        "nodeType": "Literal",
        "ruleId": "import/no-unresolved",
        "severity": 2,
      },
      Object {
        "column": 35,
        "endColumn": 45,
        "endLine": 2,
        "line": 2,
        "message": "Missing file extension for \\"../names\\"",
        "nodeType": "Literal",
        "ruleId": "import/extensions",
        "severity": 1,
      },
      Object {
        "column": 10,
        "endColumn": 22,
        "endLine": 3,
        "line": 3,
        "message": "'createPreset' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "no-unused-vars",
        "severity": 1,
      },
      Object {
        "column": 25,
        "endColumn": 46,
        "endLine": 3,
        "fix": Object {
          "range": Array [
            129,
            150,
          ],
          "text": "from '../shared'
",
        },
        "line": 3,
        "message": "Replace \`·····from·'../shared'\` with \`from·'../shared'⏎\`",
        "messageId": "replace",
        "nodeType": null,
        "ruleId": "prettier/prettier",
        "severity": 1,
      },
      Object {
        "column": 35,
        "endColumn": 46,
        "endLine": 3,
        "line": 3,
        "message": "Unable to resolve path to module '../shared'.",
        "nodeType": "Literal",
        "ruleId": "import/no-unresolved",
        "severity": 2,
      },
      Object {
        "column": 35,
        "endColumn": 46,
        "endLine": 3,
        "line": 3,
        "message": "Missing file extension for \\"../shared\\"",
        "nodeType": "Literal",
        "ruleId": "import/extensions",
        "severity": 1,
      },
    ],
    "source": "import { conditional }       from '../../shared/lib/eslint'
import { publicPresetNames } from '../names'
import { createPreset }      from '../shared'",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 8,
  },
]
`;

exports[`[Presets] Imports should arrange imports in the form of a grid: import-layout-error 1`] = `
Array [
  Object {
    "errorCount": 3,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/imports/tests/import-layout-error.js",
    "fixableErrorCount": 0,
    "fixableWarningCount": 1,
    "messages": Array [
      Object {
        "column": 10,
        "endColumn": 21,
        "endLine": 1,
        "line": 1,
        "message": "'conditional' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "no-unused-vars",
        "severity": 1,
      },
      Object {
        "column": 29,
        "endColumn": 54,
        "endLine": 1,
        "line": 1,
        "message": "Unable to resolve path to module '../../shared/lib/eslint'.",
        "nodeType": "Literal",
        "ruleId": "import/no-unresolved",
        "severity": 2,
      },
      Object {
        "column": 29,
        "endColumn": 54,
        "endLine": 1,
        "line": 1,
        "message": "Missing file extension for \\"../../shared/lib/eslint\\"",
        "nodeType": "Literal",
        "ruleId": "import/extensions",
        "severity": 1,
      },
      Object {
        "column": 10,
        "endColumn": 27,
        "endLine": 2,
        "line": 2,
        "message": "'publicPresetNames' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "no-unused-vars",
        "severity": 1,
      },
      Object {
        "column": 35,
        "endColumn": 45,
        "endLine": 2,
        "line": 2,
        "message": "Unable to resolve path to module '../names'.",
        "nodeType": "Literal",
        "ruleId": "import/no-unresolved",
        "severity": 2,
      },
      Object {
        "column": 35,
        "endColumn": 45,
        "endLine": 2,
        "line": 2,
        "message": "Missing file extension for \\"../names\\"",
        "nodeType": "Literal",
        "ruleId": "import/extensions",
        "severity": 1,
      },
      Object {
        "column": 10,
        "endColumn": 22,
        "endLine": 3,
        "line": 3,
        "message": "'createPreset' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "no-unused-vars",
        "severity": 1,
      },
      Object {
        "column": 30,
        "endColumn": 41,
        "endLine": 3,
        "line": 3,
        "message": "Unable to resolve path to module '../shared'.",
        "nodeType": "Literal",
        "ruleId": "import/no-unresolved",
        "severity": 2,
      },
      Object {
        "column": 30,
        "endColumn": 41,
        "endLine": 3,
        "line": 3,
        "message": "Missing file extension for \\"../shared\\"",
        "nodeType": "Literal",
        "ruleId": "import/extensions",
        "severity": 1,
      },
      Object {
        "column": 41,
        "endColumn": 41,
        "endLine": 3,
        "fix": Object {
          "range": Array [
            139,
            139,
          ],
          "text": "
",
        },
        "line": 3,
        "message": "Insert \`⏎\`",
        "messageId": "insert",
        "nodeType": null,
        "ruleId": "prettier/prettier",
        "severity": 1,
      },
    ],
    "source": "import { conditional } from '../../shared/lib/eslint'
import { publicPresetNames } from '../names'
import { createPreset } from '../shared'",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 7,
  },
]
`;

exports[`[Presets] Imports should match snapshots: basic-clear 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/imports/tests/basic-clear.js",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": Array [],
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 0,
  },
]
`;

exports[`[Presets] Imports should match snapshots: basic-error 1`] = `
Array [
  Object {
    "errorCount": 2,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/imports/tests/basic-error.js",
    "fixableErrorCount": 1,
    "fixableWarningCount": 0,
    "messages": Array [
      Object {
        "column": 1,
        "endColumn": 12,
        "endLine": 1,
        "fix": Object {
          "range": Array [
            0,
            3,
          ],
          "text": "let",
        },
        "line": 1,
        "message": "Unexpected var, use let or const instead.",
        "messageId": "unexpectedVar",
        "nodeType": "VariableDeclaration",
        "ruleId": "no-var",
        "severity": 2,
      },
      Object {
        "column": 1,
        "endColumn": 12,
        "endLine": 1,
        "line": 1,
        "message": "Exporting mutable 'var' binding, use 'const' instead.",
        "nodeType": "VariableDeclaration",
        "ruleId": "import/no-mutable-exports",
        "severity": 2,
      },
    ],
    "source": "var one = 1
export { one }
",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 0,
  },
]
`;

exports[`[Presets] Imports should match snapshots: basic-warn 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/imports/tests/basic-warn.js",
    "fixableErrorCount": 0,
    "fixableWarningCount": 1,
    "messages": Array [
      Object {
        "column": 1,
        "endColumn": 20,
        "endLine": 2,
        "fix": Object {
          "range": Array [
            0,
            47,
          ],
          "text": "import fs from 'fs'
import { one } from './one'",
        },
        "line": 1,
        "message": "Run autofix to sort these imports!",
        "messageId": "sort",
        "nodeType": null,
        "ruleId": "simple-import-sort/imports",
        "severity": 1,
      },
      Object {
        "column": 10,
        "endColumn": 13,
        "endLine": 1,
        "line": 1,
        "message": "'one' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "no-unused-vars",
        "severity": 1,
      },
      Object {
        "column": 8,
        "endColumn": 10,
        "endLine": 2,
        "line": 2,
        "message": "'fs' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "no-unused-vars",
        "severity": 1,
      },
    ],
    "source": "import { one } from './one'
import fs from 'fs'

export {}
",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 3,
  },
]
`;

exports[`[Presets] Imports should respect newline=false: newline-false 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/imports/tests/newline-false.js",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": Array [
      Object {
        "column": 8,
        "endColumn": 10,
        "endLine": 1,
        "line": 1,
        "message": "'fs' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "no-unused-vars",
        "severity": 1,
      },
      Object {
        "column": 10,
        "endColumn": 13,
        "endLine": 2,
        "line": 2,
        "message": "'one' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "no-unused-vars",
        "severity": 1,
      },
    ],
    "source": "import fs from 'fs'
import { one } from './one'

export {}
",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 2,
  },
]
`;

exports[`[Presets] Imports should respect newline=true: newline-true 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/imports/tests/newline-true.js",
    "fixableErrorCount": 0,
    "fixableWarningCount": 1,
    "messages": Array [
      Object {
        "column": 1,
        "endColumn": 28,
        "endLine": 2,
        "fix": Object {
          "range": Array [
            0,
            47,
          ],
          "text": "import fs from 'fs'

import { one } from './one'",
        },
        "line": 1,
        "message": "Run autofix to sort these imports!",
        "messageId": "sort",
        "nodeType": null,
        "ruleId": "simple-import-sort/imports",
        "severity": 1,
      },
      Object {
        "column": 8,
        "endColumn": 10,
        "endLine": 1,
        "line": 1,
        "message": "'fs' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "no-unused-vars",
        "severity": 1,
      },
      Object {
        "column": 10,
        "endColumn": 13,
        "endLine": 2,
        "line": 2,
        "message": "'one' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "no-unused-vars",
        "severity": 1,
      },
    ],
    "source": "import fs from 'fs'
import { one } from './one'

export {}
",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 3,
  },
]
`;

exports[`[Presets] Imports should respect paths from jsconfig.json: alias-jsconfig-clear 1`] = `
Array [
  Object {
    "errorCount": 2,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/imports/tests/alias-jsconfig-clear.js",
    "fixableErrorCount": 0,
    "fixableWarningCount": 1,
    "messages": Array [
      Object {
        "column": 1,
        "endColumn": 35,
        "endLine": 5,
        "fix": Object {
          "range": Array [
            0,
            149,
          ],
          "text": "// it doesn't work for the unknown reason
// assuming it's an environment issue
import { sum as sumB } from \\"meow\\"
import { sum as sumA } from \\"sum\\"",
        },
        "line": 1,
        "message": "Run autofix to sort these imports!",
        "messageId": "sort",
        "nodeType": null,
        "ruleId": "simple-import-sort/imports",
        "severity": 1,
      },
      Object {
        "column": 17,
        "endColumn": 21,
        "endLine": 1,
        "line": 1,
        "message": "'sumA' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "no-unused-vars",
        "severity": 1,
      },
      Object {
        "column": 29,
        "endColumn": 34,
        "endLine": 1,
        "line": 1,
        "message": "Unable to resolve path to module 'sum'.",
        "nodeType": "Literal",
        "ruleId": "import/no-unresolved",
        "severity": 2,
      },
      Object {
        "column": 17,
        "endColumn": 21,
        "endLine": 5,
        "line": 5,
        "message": "'sumB' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "no-unused-vars",
        "severity": 1,
      },
      Object {
        "column": 29,
        "endColumn": 35,
        "endLine": 5,
        "line": 5,
        "message": "Unable to resolve path to module 'meow'.",
        "nodeType": "Literal",
        "ruleId": "import/no-unresolved",
        "severity": 2,
      },
    ],
    "source": "import { sum as sumA } from \\"sum\\"

// it doesn't work for the unknown reason
// assuming it's an environment issue
import { sum as sumB } from \\"meow\\"
",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 3,
  },
]
`;

exports[`[Presets] Imports should respect paths from jsconfig.json: alias-jsconfig-error 1`] = `
Array [
  Object {
    "errorCount": 1,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/imports/tests/alias-jsconfig-error.js",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": Array [
      Object {
        "column": 10,
        "endColumn": 15,
        "endLine": 1,
        "line": 1,
        "message": "'isOdd' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "no-unused-vars",
        "severity": 1,
      },
      Object {
        "column": 23,
        "endColumn": 31,
        "endLine": 1,
        "line": 1,
        "message": "Unable to resolve path to module 'is-odd'.",
        "nodeType": "Literal",
        "ruleId": "import/no-unresolved",
        "severity": 2,
      },
    ],
    "source": "import { isOdd } from \\"is-odd\\"
",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 1,
  },
]
`;

exports[`[Presets] Imports should respect paths from tsconfig.json: alias-tsconfig-clear 1`] = `
Array [
  Object {
    "errorCount": 2,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/imports/tests/alias-tsconfig-clear.ts",
    "fixableErrorCount": 1,
    "fixableWarningCount": 1,
    "messages": Array [
      Object {
        "column": 1,
        "endColumn": 35,
        "endLine": 2,
        "fix": Object {
          "range": Array [
            0,
            68,
          ],
          "text": "import { sum as sumB } from \\"meow\\"
import { sum as sumA } from \\"sum\\"",
        },
        "line": 1,
        "message": "Run autofix to sort these imports!",
        "messageId": "sort",
        "nodeType": null,
        "ruleId": "simple-import-sort/imports",
        "severity": 1,
      },
      Object {
        "column": 17,
        "endColumn": 21,
        "endLine": 1,
        "line": 1,
        "message": "'sumA' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "@typescript-eslint/no-unused-vars",
        "severity": 1,
      },
      Object {
        "column": 29,
        "endColumn": 34,
        "endLine": 1,
        "fix": Object {
          "range": Array [
            21,
            69,
          ],
          "text": ", sum as sumB } from \\"sum\\"
",
        },
        "line": 1,
        "message": "'<PROJECT_ROOT>/src/presets/imports/tests/sum.ts' imported multiple times.",
        "nodeType": "Literal",
        "ruleId": "import/no-duplicates",
        "severity": 2,
      },
      Object {
        "column": 17,
        "endColumn": 21,
        "endLine": 2,
        "line": 2,
        "message": "'sumB' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "@typescript-eslint/no-unused-vars",
        "severity": 1,
      },
      Object {
        "column": 29,
        "endColumn": 35,
        "endLine": 2,
        "line": 2,
        "message": "'<PROJECT_ROOT>/src/presets/imports/tests/sum.ts' imported multiple times.",
        "nodeType": "Literal",
        "ruleId": "import/no-duplicates",
        "severity": 2,
      },
    ],
    "source": "import { sum as sumA } from \\"sum\\"
import { sum as sumB } from \\"meow\\"
",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 3,
  },
]
`;

exports[`[Presets] Imports should respect paths from tsconfig.json: alias-tsconfig-error 1`] = `
Array [
  Object {
    "errorCount": 1,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/imports/tests/alias-tsconfig-error.ts",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": Array [
      Object {
        "column": 10,
        "endColumn": 15,
        "endLine": 1,
        "line": 1,
        "message": "'isOdd' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "@typescript-eslint/no-unused-vars",
        "severity": 1,
      },
      Object {
        "column": 23,
        "endColumn": 31,
        "endLine": 1,
        "line": 1,
        "message": "Unable to resolve path to module 'is-odd'.",
        "nodeType": "Literal",
        "ruleId": "import/no-unresolved",
        "severity": 2,
      },
    ],
    "source": "import { isOdd } from \\"is-odd\\"
",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 1,
  },
]
`;
